### PR TITLE
make Object::Clone() more or less work for arrays

### PR DESCRIPTION
**Pull request by tbsaunde**
@ehsan or @bzbarsky can you take a quick look at the sanity of the property coppying stuff here?

### CODE REVIEW

File: deps/spidershim/src/v8object.cc [hunk 1]
```diff
@@ -454,6 +454,38 @@ Local<Object> Object::Clone() {
...
+    }
+
+    JS::RootedObject copy(cx, JS_NewArrayObject(cx, length));
+    if (!copy) {
+      return Local<Object>();
+    }
+
+    JS::RootedValue val(cx);
+    for (uint32_t i = 0; i < length; i++) {
+      if (!JS_GetElement(cx, thisObj, i, &val)) {
```
Comment (bzbarsky) [inline deps/spidershim/src/v8object.cc:475]: So... this won't do the "right thing" for accessor properties.  For that matter, it won't copy non-integer-named properties.  Or preserve holes.  I'm not sure what semantics you want here, though.

File: deps/spidershim/src/v8object.cc [hunk 2]
```diff
@@ -454,6 +454,38 @@ Local<Object> Object::Clone() {
...
+        return Local<Object>();
+      }
+
+      if (!JS_SetElement(cx, copy, i, val)) {
```
Comment (bzbarsky) [inline deps/spidershim/src/v8object.cc]: You want JS_DefineElement.  And again, not clear what you want in terms of whether property enumerability, configurability, etc is preserved.

**Comment by tbsaunde** [comment]
On Thu, Nov 17, 2016 at 01:11:17PM -0800, Boris Zbarsky wrote:

> @bzbarsky commented on this pull request.
> 
> ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
> 
> In deps/spidershim/src/v8object.cc:
> 
> > -  }
> >   +
> > -  if (isArray) {
> > -    uint32_t length;
> > -    if (!JS_GetArrayLength(cx, thisObj, &length)) {
> > -      return Local<Object>();
> > -    }
> >   +
> > -    JS::RootedObject copy(cx, JS_NewArrayObject(cx, length));
> > -    if (!copy) {
> > -      return Local<Object>();
> > -    }
> >   +
> > -    JS::RootedValue val(cx);
> > -    for (uint32_t i = 0; i < length; i++) {
> > -      if (!JS_GetElement(cx, thisObj, i, &val)) {
> 
> So... this won't do the "right thing" for accessor properties. For that matter,
> it won't copy non-integer-named properties. Or preserve holes. I'm not sure
> what semantics you want here, though.

So,  the one consumer of Clone() looks like this.

array = val.Clone();
for (i = 0; i < array->Length(); i++)
  array[i] = array[i]->ToString();

it then proceeds to copy the strings out of js strings and throw the
array away so I don't understand why it makes a clone at all, but
anyway.

It seems like non integer name properties don't matter, however holes do
matter.

for accessor properties I guess JS_GetElement calls the getter and
returns its return value?  That's probably fine here though I suspect
that the correct thing to do would be make a shallow copy of the
descriptor and then the ToString call in this specific case would deal
with the rest.

> ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
> 
> In deps/spidershim/src/v8object.cc:
> 
> > -    if (!JS_GetArrayLength(cx, thisObj, &length)) {
> > -      return Local<Object>();
> > -    }
> >   +
> > -    JS::RootedObject copy(cx, JS_NewArrayObject(cx, length));
> > -    if (!copy) {
> > -      return Local<Object>();
> > -    }
> >   +
> > -    JS::RootedValue val(cx);
> > -    for (uint32_t i = 0; i < length; i++) {
> > -      if (!JS_GetElement(cx, thisObj, i, &val)) {
> > -        return Local<Object>();
> > -      }
> >   +
> > -      if (!JS_SetElement(cx, copy, i, val)) {
> 
> You want JS_DefineElement. And again, not clear what you want in terms of
> whether property enumerability, configurability, etc is preserved.

ah ok.

Looking at how V8 does this stuff it looks like an object has an array
of Object\* for properties and another for elements.  When they clone
they just copy those pointers, so I believe that they are preserved,
though of course if we only care about the one user it would seem that
doesn't really matter.

**Comment by bzbarsky** [comment]
> It seems like non integer name properties don't matter, however holes do matter.

How do they matter?  It's just going 0 to length, so treating holes as undefined-valued properties, not as holes...

> for accessor properties I guess JS_GetElement calls the getter and returns its return value?

That's correct.

**Comment by tbsaunde** [comment]
On Thu, Nov 17, 2016 at 03:08:17PM -0800, Boris Zbarsky wrote:

> ```
> It seems like non integer name properties don't matter, however holes do
> matter.
> ```
> 
> How do they matter? It's just going 0 to length, so treating holes as
> undefined-valued properties, not as holes...

I'm not sure I understand the difference between holes and elements where
the value is undefined.

> ```
> for accessor properties I guess JS_GetElement calls the getter and returns
> its return value?
> ```
> 
> That's correct.

ok, so in theory we should probably copy the property descriptor or
something, but for the only existing user it doesn't matter.

**Comment by bzbarsky** [comment]
> I'm not sure I understand the difference between holes and elements where the value is undefined.

The difference is whether gets fall through to Array.prototype, whether getOwnPropertyDescriptor claims there is a property, whether indexOf/lastIndexOf(undefined) return -1, whether map/reduce/reduceRight/reverse/some() see the undefined value,  and probably some other ones I'm forgetting.

**Comment by mykmelez** [comment]
@tbsaunde Work seems to have stalled here. Is this worth keeping open, or shall we close it as incomplete and re-open if work picks back up?

**Comment by mykmelez** [comment]
I'm closing this pull request, as development has stalled on the project, so I'm going to archive it, in the interest of representing its status accurately.

**Pull request closed without merging**